File: ios/Plugin/Capacitor3KakaoLogin.swift
import Foundation
import Capacitor

import KakaoSDKUser
import KakaoSDKCommon
import KakaoSDKShare
import KakaoSDKTemplate


@objc public class Capacitor3KakaoLogin: NSObject {
    @objc public func kakaoLogin(_ call: CAPPluginCall) -> Void {
        
        // 카카오톡 설치 여부 확인
        if (UserApi.isKakaoTalkLoginAvailable()) {
            UserApi.shared.loginWithKakaoTalk {(oauthToken, error) in
                if let error = error {
                    print(error)
                    call.reject("error")
                }
                else {

                    call.resolve([
                        "value": oauthToken?.accessToken ?? ""
                    ])
                    
                }
            }
        }
        else{
            
            UserApi.shared.loginWithKakaoAccount {(oauthToken, error) in
                    if let error = error {
                        print(error)
                        call.reject("error")
                    }
                    else {

                        call.resolve([
                            "value": oauthToken?.accessToken ?? ""
                        ])
                    }
                }
            
        }
    }
    
    @objc public func kakaoLogout(_ call: CAPPluginCall) -> Void {

        UserApi.shared.logout {(error) in
            if let error = error {
                print(error)
                call.reject("error")
            }
            else {

                call.resolve([
                    "value": "done"
                ])
            }
        }
    }
    
    
    
    @objc public func kakaoUnlink(_ call: CAPPluginCall) -> Void {

        UserApi.shared.unlink {(error) in
            if let error = error {
                print(error)
                call.reject("error")
            }
            else {

                call.resolve([
                    "value": "done"
                ])
            }
        }
    }
    
    
    @objc public func sendLinkFeed(_ call: CAPPluginCall) -> Void {

        let title = call.getString("title") ?? ""
        let description = call.getString("description") ?? ""
        let image_url = call.getString("image_url") ?? ""
        let image_link_url = call.getString("image_link_url") ?? ""
        let button_title = call.getString("button_title") ?? ""

        
        
        let link = Link(webUrl: URL(string:image_link_url),
                        mobileWebUrl: URL(string:image_link_url))

        let button = Button(title: button_title, link: link)
       
        let content = Content(title: title,
                                imageUrl: URL(string:image_url)!,
                                description: description,
                                link: link)
        let feedTemplate = FeedTemplate(content: content, social: nil, buttons: [button])

//        //메시지 템플릿 encode
//        if let feedTemplateJsonData = (try? SdkJSONEncoder.custom.encode(feedTemplate)) {
//
//        //생성한 메시지 템플릿 객체를 jsonObject로 변환
//            if let templateJsonObject = SdkUtils.toJsonObject(feedTemplateJsonData) {
//                ShareApi.shared.shareDefault(templateObject:templateJsonObject) {(linkResult, error) in
//                    if let error = error {
//                        print(error)
//                        call.reject("error")
//                    }
//                    else {
//
//                        //do something
//                        guard let linkResult = linkResult else { return }
//                        UIApplication.shared.open(linkResult.url, options: [:], completionHandler: nil)
//
//                        call.resolve([
//                            "value": "done"
//                        ])
//                    }
//                }
//            }
//
//        }
        
        
        if let feedTemplateJsonData = (try? SdkJSONEncoder.custom.encode(feedTemplate)) {

        //jsonObject
            if let templateJsonObject = SdkUtils.toJsonObject(feedTemplateJsonData) {
                ShareApi.shared.shareDefault(templateObject:templateJsonObject) { [weak self] (sharingResult, error) in
                    if let error = error {
                        print(error)
                        call.reject("error")
                    }
                    else {
                        //do something
                        guard let sharingResult = sharingResult else { return }
                        UIApplication.shared.open(sharingResult.url, options: [:], completionHandler: nil)
                    
                        
                        call.resolve([
                            "value": "done"
                        ])
                    }
                }
            }
        }
        
        
        
    }
}

File: android/src/main/java/com/fumi/capacitor3_kakao_login/Capacitor3KakaoLogin.java
package com.fumi.capacitor3_kakao_login;

import android.content.Context;
import android.util.Log;
import androidx.appcompat.app.AppCompatActivity;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.kakao.sdk.common.KakaoSdk;
import com.kakao.sdk.share.ShareClient;
import com.kakao.sdk.template.model.Button;
import com.kakao.sdk.template.model.Content;
import com.kakao.sdk.template.model.FeedTemplate;
import com.kakao.sdk.template.model.Link;
import com.kakao.sdk.user.UserApiClient;
import java.util.ArrayList;

public class Capacitor3KakaoLogin {

    private static final String TAG = "Capacitor3KakaoLogin";
    AppCompatActivity activity;

    public Capacitor3KakaoLogin(AppCompatActivity activity) {
        this.activity = activity;
    }

    public void kakaoLogin(PluginCall call) {
        if (UserApiClient.getInstance().isKakaoTalkLoginAvailable(this.activity)) {
            UserApiClient
                    .getInstance()
                    .loginWithKakaoTalk(
                            this.activity,
                            (oAuthToken, error) -> {
                                if (error != null) {
                                    Log.e(TAG, "로그인 실패", error);
                                    call.reject(error.toString());
                                } else if (oAuthToken != null) {
                                    Log.i(TAG, "로그인 성공(토큰) : " + oAuthToken.getAccessToken());
                                    JSObject ret = new JSObject();
                                    ret.put("value", oAuthToken.getAccessToken());
                                    call.resolve(ret);
                                } else {
                                    call.reject("no_data");
                                }
                                return null;
                            }
                    );
        } else {
            UserApiClient
                    .getInstance()
                    .loginWithKakaoAccount(
                            this.activity,
                            (oAuthToken, error) -> {
                                if (error != null) {
                                    Log.e(TAG, "로그인 실패", error);
                                    call.reject(error.toString());
                                } else if (oAuthToken != null) {
                                    Log.i(TAG, "로그인 성공(토큰) : " + oAuthToken.getAccessToken());
                                    JSObject ret = new JSObject();
                                    ret.put("value", oAuthToken.getAccessToken());
                                    call.resolve(ret);
                                } else {
                                    call.reject("no_data");
                                }
                                return null;
                            }
                    );
        }
    }

    public void kakaoLogout(PluginCall call) {
        UserApiClient
                .getInstance()
                .logout(
                        error -> {
                            JSObject ret = new JSObject();
                            ret.put("value", "done");
                            call.resolve(ret);
                            return null;
                        }
                );
    }

    public void kakaoUnlink(PluginCall call) {
        UserApiClient
                .getInstance()
                .unlink(
                        error -> {
                            JSObject ret = new JSObject();
                            ret.put("value", "done");
                            call.resolve(ret);
                            return null;
                        }
                );
    }

    public void sendLinkFeed(PluginCall call) {
        Link link = new Link(call.getString("image_link_url"), call.getString("image_link_url"), null, null);
        Content content = new Content(call.getString("title"), call.getString("image_url"), link, call.getString("description"));
        ArrayList<Button> buttons = new ArrayList<>();
        buttons.add(new Button(call.getString("button_title"), link));
        FeedTemplate feed = new FeedTemplate(content, null, null, buttons);
        ShareClient
                .getInstance()
                .shareDefault(
                        this.activity,
                        feed,
                        (linkResult, error) -> {
                            if (error != null) {} else if (linkResult != null) {
                                this.activity.startActivity(linkResult.getIntent());
                            }

                            JSObject ret = new JSObject();
                            ret.put("value", "done");
                            call.resolve(ret);
                            return null;
                        }
                );
    }
}
